sagemaker_shim/models.py

```diff
@@ -288,61 +354,27 @@ def proc_env(self) -> dict[str, str]:
 
         return env
 
-    @staticmethod
-    def _get_user_info(id_or_name: str) -> pwd.struct_passwd | None:
-        if id_or_name == "":
-            return None
-
-        try:
-            return pwd.getpwnam(id_or_name)
-        except (KeyError, AttributeError):
-            try:
-                return pwd.getpwuid(int(id_or_name))
-            except (KeyError, ValueError, AttributeError) as error:
-                raise RuntimeError(f"User {id_or_name} not found") from error
-
-    @staticmethod
-    def _get_group_info(id_or_name: str) -> grp.struct_group | None:
-        if id_or_name == "":
-            return None
-
-        try:
-            return grp.getgrnam(id_or_name)
-        except (KeyError, AttributeError):
-            try:
-                return grp.getgrgid(int(id_or_name))
-            except (KeyError, ValueError, AttributeError) as error:
-                raise RuntimeError(f"Group {id_or_name} not found") from error
-
     @cached_property
-    def proc_user(self) -> UserGroup:
+    def proc_user(self) -> UserInfo:
+        if self.user == "":
+            return UserInfo(uid=None, gid=None, home=None, groups=[])
+
         match = re.fullmatch(
             r"^(?P<user>[0-9a-zA-Z]*):?(?P<group>[0-9a-zA-Z]*)$", self.user
         )
 
         if match:
-            user = self._get_user_info(id_or_name=match.group("user"))
-            group = self._get_group_info(id_or_name=match.group("group"))
-
-            if user is None:
-                uid = None
-                home = None
-            else:
-                uid = user.pw_uid
-                home = user.pw_dir
-
-            if group is None:
-                if user is None:
-                    gid = None
-                else:
-                    # Switch to the users primary group
-                    gid = user.pw_gid
-            else:
-                gid = group.gr_gid
-
-            return UserGroup(uid=uid, gid=gid, home=home)
+            info = _get_user_info(id_or_name=match.group("user"))
+            gid = _get_group_id(id_or_name=match.group("group"))
+
+            return UserInfo(
+                uid=info.uid,
+                gid=info.gid if gid is None else gid,
```

Comment:
Here we set an explicit `gid` for the process (`group=gid` in Popen). Does that `gid` also need to be added to `groups` if it is not there already?

Reply:
No, not that I know. The supplemental groups are handled somehow differently from the uid/gid. Though I do not quite know how those work either. I quickly tried to find out now by inspecting the `/proc/[prog-id]` directory, but I do not think that the supplemental groups are listed there. The gid is... no clue. The supplemental/extra groups must be resolved using some other mechanism...

Reply:
Well, digging further makes me question my earlier answer:

https://www.man7.org/linux/man-pages/man2/getgroups.2.html

> It is unspecified whether the effective group ID of the calling process is included in the returned list.  (Thus, an application should also call [getegid(2)](https://www.man7.org/linux/man-pages/man2/getegid.2.html) and add or remove the resulting value.)

Great! So maybe we should add it just-in-case?

Reply:
Done!